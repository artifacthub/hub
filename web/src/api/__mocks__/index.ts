export const API = {
  getPackage: jest.fn(),
  toggleStar: jest.fn(),
  getStars: jest.fn(),
  searchPackages: jest.fn(),
  getStats: jest.fn(),
  getRandomPackages: jest.fn(),
  register: jest.fn(),
  verifyEmail: jest.fn(),
  login: jest.fn(),
  logout: jest.fn(),
  getUserProfile: jest.fn(),
  getAllRepositories: jest.fn(),
  getRepositories: jest.fn(),
  addRepository: jest.fn(),
  deleteRepository: jest.fn(),
  updateRepository: jest.fn(),
  transferRepository: jest.fn(),
  claimRepositoryOwnership: jest.fn(),
  checkAvailability: jest.fn(),
  getUserOrganizations: jest.fn(),
  getOrganization: jest.fn(),
  addOrganization: jest.fn(),
  updateOrganization: jest.fn(),
  deleteOrganization: jest.fn(),
  getOrganizationMembers: jest.fn(),
  addOrganizationMember: jest.fn(),
  deleteOrganizationMember: jest.fn(),
  confirmOrganizationMembership: jest.fn(),
  getStarredByUser: jest.fn(),
  updateUserProfile: jest.fn(),
  updatePassword: jest.fn(),
  saveImage: jest.fn(),
  getPackageSubscriptions: jest.fn(),
  addSubscription: jest.fn(),
  deleteSubscription: jest.fn(),
  getUserSubscriptions: jest.fn(),
  getWebhooks: jest.fn(),
  getWebhook: jest.fn(),
  addWebhook: jest.fn(),
  deleteWebhook: jest.fn(),
  updateWebhook: jest.fn(),
  triggerWebhookTest: jest.fn(),
  getAPIKeys: jest.fn(),
  getAPIKey: jest.fn(),
  addAPIKey: jest.fn(),
  updateAPIKey: jest.fn(),
  deleteAPIKey: jest.fn(),
  getOptOutList: jest.fn(),
  addOptOut: jest.fn(),
  deleteOptOut: jest.fn(),
  getAuthorizationPolicy: jest.fn(),
  updateAuthorizationPolicy: jest.fn(),
  getUserAllowedActions: jest.fn(),
  getSnapshotSecurityReport: jest.fn(),
  getValuesSchema: jest.fn(),
  getChangelog: jest.fn(),
  triggerTestInRegoPlayground: jest.fn(),
};
