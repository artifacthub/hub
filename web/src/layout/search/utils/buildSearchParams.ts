import { Filters } from '../../../types';
import isNull from 'lodash/isNull';

const SPECIAL_KEYS = ['text', 'page'];

interface F {
  [key: string]: string[];
}

export default (query: string): Filters => {
  const p = new URLSearchParams(query);
  let f: F = {};

  p.forEach((value, key) => {
    if (!SPECIAL_KEYS.includes(key)) {
      const values = f[key] || [];
      values.push(value);
      f[key] = values;
    }
  });

  return {
    text: p.has('text') ? p.get('text')! : undefined,
    f: { ...f },
    pageNumber: p.has('page') && !isNull(p.get('page')) ? parseInt(p.get('page')!) : 1,
  };
}
