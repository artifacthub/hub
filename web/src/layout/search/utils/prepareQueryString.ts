import isUndefined from 'lodash/isUndefined';
import { Filters } from '../../../types';

export default (filters: Filters): string => {
  const p = new URLSearchParams();
  p.set('page', filters.pageNumber.toString());
  if (!isUndefined(filters.text)) {
    p.set('text', filters.text);
  }
  if (!isUndefined(filters.f)) {
    Object.keys(filters.f).forEach((filterId: string) => {
      return filters.f[filterId].forEach((id: string) => {
        p.append(filterId, id);
      });
    });
  }
  const result = p.toString();
  return result === '' ? '' : `?${result}`;
}
