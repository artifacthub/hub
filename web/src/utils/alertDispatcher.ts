import isUndefined from 'lodash/isUndefined';

import { Alert } from '../types';

const DEFUALT_TIME = 5 * 1000; //5s

export interface AlertUpdatesHandler {
  updateAlertWrapper(alert: Alert | null): void;
}

export class AlertDispatcher {
  private updatesHandler?: AlertUpdatesHandler;
  private visibleAlert: Alert | null = null;
  private timeout?: NodeJS.Timeout;

  public subscribe(updatesHandler: AlertUpdatesHandler) {
    this.updatesHandler = updatesHandler;
  }

  public postAlert(alert: Alert | null) {
    this.visibleAlert = alert;
    this.dismissAlert();
    if (this.updatesHandler) {
      this.updatesHandler.updateAlertWrapper(this.visibleAlert);
    }
  }

  private dismissAlert() {
    this.clearTimeout();
    if (this.visibleAlert && (isUndefined(this.visibleAlert.autoClose) || this.visibleAlert.autoClose)) {
      this.timeout = setTimeout(() => {
        this.postAlert(null);
      }, this.visibleAlert.dismissOn || DEFUALT_TIME);
    }
  }

  private clearTimeout() {
    if (this.timeout) {
      clearTimeout(this.timeout);
    }
  }
}

const alertDispatcher = new AlertDispatcher();
export default alertDispatcher;
