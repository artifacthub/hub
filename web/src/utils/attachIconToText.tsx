import classnames from 'classnames';
import { isNull } from 'lodash';
import React, { useEffect, useRef, useState } from 'react';

interface Props {
  lineHeight?: number;
  className?: string;
  text: string;
  icon: JSX.Element;
  isVisible: boolean;
}

const DEFAULT_LINE_HEIGHT = 21;

export default (props: Props): JSX.Element => {
  const lineHeight: number = props.lineHeight || DEFAULT_LINE_HEIGHT;
  const ref = useRef<HTMLDivElement | null>(null);
  const icon = useRef<HTMLDivElement | null>(null);
  const lastCharacter = useRef<HTMLDivElement | null>(null);
  const [reducedFontSize, setreducedFontSize] = useState<boolean>(false);

  useEffect(() => {
    const calculateFontSize = () => {
      const { offsetTop } = ref.current!;
      const lineOfLastCharacter = Math.floor((lastCharacter.current!.offsetTop - offsetTop) / lineHeight);
      const lineOfIcon = Math.floor((icon.current!.offsetTop - offsetTop) / lineHeight);
      if (lineOfIcon !== lineOfLastCharacter) {
        setreducedFontSize(true);
      }
    };

    if (!isNull(ref.current) && props.isVisible) {
      calculateFontSize();
    }
  }, [props.isVisible, lineHeight]);

  return (
    <div
      ref={ref}
      className={props.className}
      style={{
        fontSize: reducedFontSize ? '90%' : 'inherit',
        lineHeight: `${lineHeight}px`,
      }}
    >
      {props.text.split('').map((ch: string, index: number) => {
        if (index + 1 === props.text.length) {
          return (
            <div className="d-inline" key={`ch_${index}`}>
              {ch}
            </div>
          );
        } else {
          return (
            <div className="d-inline" ref={lastCharacter} key={`ch_${index}`}>
              {ch}
            </div>
          );
        }
      })}
      <div ref={icon} className={classnames('d-inline-block', { 'position-absolute': reducedFontSize })}>
        {props.icon}
      </div>
    </div>
  );
};
