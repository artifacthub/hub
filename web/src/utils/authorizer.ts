import { isUndefined } from 'lodash';

import { API } from '../api';
import { AuthorizerAction, AuthorizerInput } from '../types';

export class Authorizer {
  private allowedActions: (AuthorizerAction | string)[] = ['all'];
  private selectedOrg?: string;
  private gettingActions: boolean = false;
  private pendingActions: (() => void)[] = [];

  public init(selectedOrg?: string) {
    this.selectedOrg = selectedOrg;
    if (selectedOrg) {
      this.getAllowedActionsList();
    }
  }

  public updateCtx(selectedOrg?: string) {
    if (isUndefined(selectedOrg)) {
      this.selectedOrg = undefined;
      this.allowedActions = ['all'];
    } else if (this.selectedOrg !== selectedOrg) {
      this.selectedOrg = selectedOrg;
      this.getAllowedActionsList();
    }
  }

  public check(input: AuthorizerInput): boolean {
    if (input.organizationName && input.organizationName !== this.selectedOrg) {
      this.selectedOrg = input.organizationName;
      this.getAllowedActionsList();
      return true;
    } else {
      if (this.gettingActions && input.onCompletion && input.organizationName) {
        this.pendingActions.push(input.onCompletion);
        return true;
      } else {
        if (isUndefined(this.selectedOrg) || this.allowedActions.includes('all')) {
          return true;
        } else {
          return this.allowedActions.includes(input.action);
        }
      }
    }
  }

  public getAllowedActionsList(onCompletion?: () => void) {
    const getUserAllowedActions = async () => {
      try {
        this.gettingActions = true;
        this.allowedActions = await API.getUserAllowedActions(this.selectedOrg!);
        this.gettingActions = false;
        if (onCompletion) {
          onCompletion();
        }
        if (this.pendingActions.length > 0) {
          this.pendingActions.forEach((action: () => void) => action());
          this.pendingActions = [];
        }
      } catch (err) {
        this.allowedActions = ['all'];
        this.gettingActions = false;
      }
    };

    if (this.selectedOrg) {
      getUserAllowedActions();
    }
  }
}

const authorizer = new Authorizer();
export default authorizer;
