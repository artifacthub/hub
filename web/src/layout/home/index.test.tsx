import { render, waitFor } from '@testing-library/react';
import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import { mocked } from 'ts-jest/utils';

import { API } from '../../api';
import { ErrorKind, Stats } from '../../types';
import HomeView from './index';
jest.mock('../../api');
jest.mock('./SearchTip', () => () => <div />);
jest.mock('../common/SampleQueries', () => () => <div />);
jest.mock('./RandomPackages', () => () => <div />);

const getMockStats = (fixtureId: string): Stats => {
  return require(`./__fixtures__/index/${fixtureId}.json`) as Stats;
};

const defaultProps = {
  isSearching: true,
  onOauthFailed: false,
};

describe('Home index', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', async () => {
    const mockStats = getMockStats('1');
    mocked(API).getStats.mockResolvedValue(mockStats);

    const result = render(
      <Router>
        <HomeView {...defaultProps} />
      </Router>
    );

    await waitFor(() => {
      expect(API.getStats).toHaveBeenCalledTimes(1);
      expect(result.asFragment()).toMatchSnapshot();
    });
  });

  describe('Render', () => {
    it('renders component', async () => {
      const mockStats = getMockStats('2');
      mocked(API).getStats.mockResolvedValue(mockStats);

      render(
        <Router>
          <HomeView {...defaultProps} />
        </Router>
      );

      await waitFor(() => {
        expect(API.getStats).toHaveBeenCalledTimes(1);
      });
    });

    it('renders dash symbol when results are 0', async () => {
      const mockStats = getMockStats('4');
      mocked(API).getStats.mockResolvedValue(mockStats);

      const props = {
        ...defaultProps,
        isSearching: true,
      };

      const { getAllByText } = render(
        <Router>
          <HomeView {...props} />
        </Router>
      );

      const emptyStats = await waitFor(() => getAllByText('-'));
      expect(emptyStats).toHaveLength(2);
    });

    it('renders dash symbol when getStats call fails', async () => {
      mocked(API).getStats.mockRejectedValue({ kind: ErrorKind.Other });

      const props = {
        ...defaultProps,
        isSearching: true,
      };

      const { getAllByText } = render(
        <Router>
          <HomeView {...props} />
        </Router>
      );

      await waitFor(() => expect(API.getStats).toHaveBeenCalledTimes(1));
      expect(getAllByText('-')).toHaveLength(2);
    });

    it('renders project definition', async () => {
      const mockStats = getMockStats('5');
      mocked(API).getStats.mockResolvedValue(mockStats);

      const { getByRole, getByText } = render(
        <Router>
          <HomeView {...defaultProps} />
        </Router>
      );

      const heading = await waitFor(() => getByRole('heading'));

      expect(heading).toBeInTheDocument();
      expect(getByText('Find, install and publish')).toBeInTheDocument();
      expect(getByText('Kubernetes packages')).toBeInTheDocument();
    });
  });

  describe('External links', () => {
    it('renders proper links', async () => {
      const mockStats = getMockStats('5');
      mocked(API).getStats.mockResolvedValue(mockStats);

      const { getAllByRole } = render(
        <Router>
          <HomeView {...defaultProps} />
        </Router>
      );

      await waitFor(() => expect(API.getStats).toHaveBeenCalledTimes(1));

      const links = getAllByRole('button');
      expect(links).toHaveLength(16);

      expect(links[2]).toHaveProperty('href', 'https://github.com/cncf/hub');
      expect(links[3]).toHaveProperty('href', 'https://cloud-native.slack.com/channels/artifact-hub');
      expect(links[4]).toHaveProperty('href', 'https://twitter.com/cncfartifacthub');

      // Docs link
      expect(links[5]).toHaveProperty('href', 'http://localhost/docs/topics/repositories');

      // Packages
      expect(links[6]).toHaveProperty('href', 'https://helm.sh/');
      expect(links[7]).toHaveProperty('href', 'https://falco.org/');
      expect(links[8]).toHaveProperty('href', 'https://www.openpolicyagent.org/');
      expect(links[9]).toHaveProperty('href', 'https://github.com/operator-framework');
      expect(links[10]).toHaveProperty('href', 'https://tinkerbell.org/');
      expect(links[11]).toHaveProperty('href', 'https://krew.sigs.k8s.io/');
      expect(links[12]).toHaveProperty('href', 'https://tekton.dev/');
      expect(links[13]).toHaveProperty('href', 'https://keda.sh/');
      expect(links[14]).toHaveProperty('href', 'https://coredns.io/');

      expect(links[15]).toHaveProperty('href', 'https://www.cncf.io/sandbox-projects/');
    });
  });
});
