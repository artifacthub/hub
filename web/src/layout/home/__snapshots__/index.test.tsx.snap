// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home index creates snapshot 1`] = `
<DocumentFragment>
  <div
    class="d-flex flex-column flex-grow-1 home home"
  >
    <div
      class="jumbotron mb-0 text-center jumbotron"
    >
      <div
        class="display-4 text-center font-weight-light d-block d-xxl-flex justify-content-center mainTitle"
        role="heading"
      >
        <div>
          Find, install and publish
        </div>
        <div
          class="secondLine"
        >
          Kubernetes packages
        </div>
      </div>
      <div
        class="mt-5 text-center"
      >
        <div
          class="position-relative m-auto w-50 search"
        >
          <div
            class="d-flex align-items-strecht overflow-hidden searchBar searchBar big"
          >
            <div
              class="d-none d-sm-flex align-items-center iconWrapper"
              data-testid="searchBarIcon"
            >
              <svg
                fill="none"
                height="1em"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  cx="11"
                  cy="11"
                  r="8"
                />
                <line
                  x1="21"
                  x2="16.65"
                  y1="21"
                  y2="16.65"
                />
              </svg>
            </div>
            <input
              aria-label="Search"
              autocapitalize="none"
              autocomplete="off"
              autocorrect="off"
              class="flex-grow-1 pl-sm-0 pl-3 input"
              disabled=""
              placeholder="Search packages"
              spellcheck="false"
              type="text"
              value=""
            />
            <div
              class="position-absolute text-secondary loading bigLoading"
            >
              <span
                class="spinner-border spinner-border-sm"
                data-testid="searchBarSpinning"
              />
            </div>
            <button
              aria-label="Close"
              class="close inputClean invisible"
              data-testid="cleanBtn"
              type="button"
            >
              <span
                aria-hidden="true"
              >
                Ã—
              </span>
            </button>
            <div
              class="position-absolute text-dark tipIcon bigTipIcon"
            >
              <button
                class="btn btn-link p-2 text-light btn-lg"
                data-testid="openSearchTipsBtn"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 512 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 448c-110.532 0-200-89.431-200-200 0-110.495 89.472-200 200-200 110.491 0 200 89.471 200 200 0 110.53-89.431 200-200 200zm107.244-255.2c0 67.052-72.421 68.084-72.421 92.863V300c0 6.627-5.373 12-12 12h-45.647c-6.627 0-12-5.373-12-12v-8.659c0-35.745 27.1-50.034 47.579-61.516 17.561-9.845 28.324-16.541 28.324-29.579 0-17.246-21.999-28.693-39.784-28.693-23.189 0-33.894 10.977-48.942 29.969-4.057 5.12-11.46 6.071-16.666 2.124l-27.824-21.098c-5.107-3.872-6.251-11.066-2.644-16.363C184.846 131.491 214.94 112 261.794 112c49.071 0 101.45 38.304 101.45 88.8zM298 368c0 23.159-18.841 42-42 42s-42-18.841-42-42 18.841-42 42-42 42 18.841 42 42z"
                  />
                </svg>
              </button>
              <div>
                <div
                  class="modal modal"
                  role="dialog"
                >
                  <div
                    class="modal-dialog modal-lg modal-dialog-centered modal-dialog-scrollable"
                  >
                    <div
                      class="modal-content content"
                    >
                      <div
                        class="modal-body p-4 h-100 d-flex flex-column"
                      >
                        <div>
                          <div
                            class="alertWrapper"
                            data-testid="alertWrapper"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div />
        <div
          class="d-inline-block d-md-none text-center mt-3"
        >
          - or -
          <a
            class="btn btn-link textLight font-weight-bold py-0 pb-1 pl-1"
            href="/packages/search"
          >
            <u>
              browse all packages
            </u>
          </a>
        </div>
        <div
          class="d-none d-md-inline-block text-center mt-5"
        >
          You can also 
          <a
            class="btn btn-link textLight font-weight-bold py-0 pb-1 pl-1"
            href="/packages/search"
          >
            <u>
              browse all packages
            </u>
          </a>
           - or - 
          <span
            class="ml-3"
          >
            try one of the sample queries:
          </span>
        </div>
        <div
          class="d-none d-md-flex flex-row align-items-end justify-content-center flex-wrap"
        >
          <div />
        </div>
      </div>
      <div
        class="d-flex align-items-center justify-content-center mt-4 mt-md-5"
      >
        <div
          class="text-center counterWrapper"
        >
          <div
            class="h3 counter"
          >
            <div
              class="spinner-grow text-primary"
              role="status"
            />
          </div>
          <small
            class="text-uppercase"
          >
            packages
          </small>
        </div>
        <div
          class="mx-3 mx-md-5 separator separator"
        />
        <div
          class="text-center counterWrapper"
        >
          <div
            class="h3 counter"
          >
            <div
              class="spinner-grow text-primary"
              role="status"
            />
          </div>
          <small
            class="text-uppercase"
          >
            releases
          </small>
        </div>
      </div>
      <div
        class="text-center h5 my-4 mt-md-5 legend"
      >
        Artifact Hub is an 
        <b>
          Open Source
        </b>
         project
      </div>
      <div
        class="d-flex flex-row align-items-center justify-content-center flex-wrap"
      >
        <a
          class="link btn btn-secondary mb-4 mb-md-2 socialBtn"
          href="https://github.com/cncf/hub"
          rel="noopener noreferrer"
          role="button"
          target="_blank"
        >
          <div
            class="d-flex align-items-center justify-content-center"
          >
            <svg
              class="mr-2"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 496 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"
              />
            </svg>
            GitHub
          </div>
        </a>
        <a
          class="link btn btn-secondary ml-2 ml-md-3 mb-4 mb-md-2 socialBtn"
          href="https://cloud-native.slack.com/channels/artifact-hub"
          rel="noopener noreferrer"
          role="button"
          target="_blank"
        >
          <div
            class="d-flex align-items-center justify-content-center"
          >
            <svg
              class="mr-2"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M94.12 315.1c0 25.9-21.16 47.06-47.06 47.06S0 341 0 315.1c0-25.9 21.16-47.06 47.06-47.06h47.06v47.06zm23.72 0c0-25.9 21.16-47.06 47.06-47.06s47.06 21.16 47.06 47.06v117.84c0 25.9-21.16 47.06-47.06 47.06s-47.06-21.16-47.06-47.06V315.1zm47.06-188.98c-25.9 0-47.06-21.16-47.06-47.06S139 32 164.9 32s47.06 21.16 47.06 47.06v47.06H164.9zm0 23.72c25.9 0 47.06 21.16 47.06 47.06s-21.16 47.06-47.06 47.06H47.06C21.16 243.96 0 222.8 0 196.9s21.16-47.06 47.06-47.06H164.9zm188.98 47.06c0-25.9 21.16-47.06 47.06-47.06 25.9 0 47.06 21.16 47.06 47.06s-21.16 47.06-47.06 47.06h-47.06V196.9zm-23.72 0c0 25.9-21.16 47.06-47.06 47.06-25.9 0-47.06-21.16-47.06-47.06V79.06c0-25.9 21.16-47.06 47.06-47.06 25.9 0 47.06 21.16 47.06 47.06V196.9zM283.1 385.88c25.9 0 47.06 21.16 47.06 47.06 0 25.9-21.16 47.06-47.06 47.06-25.9 0-47.06-21.16-47.06-47.06v-47.06h47.06zm0-23.72c-25.9 0-47.06-21.16-47.06-47.06 0-25.9 21.16-47.06 47.06-47.06h117.84c25.9 0 47.06 21.16 47.06 47.06 0 25.9-21.16 47.06-47.06 47.06H283.1z"
              />
            </svg>
            Slack
          </div>
        </a>
        <a
          class="link btn btn-secondary ml-2 ml-md-3 mb-4 mb-md-2 socialBtn"
          href="https://twitter.com/cncfartifacthub"
          rel="noopener noreferrer"
          role="button"
          target="_blank"
        >
          <div
            class="d-flex align-items-center justify-content-center"
          >
            <svg
              class="mr-2"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"
              />
            </svg>
            Twitter
          </div>
        </a>
      </div>
      <div
        class="text-center mx-3 mt-md-4 mb-4 repoGuideText"
      >
        Please see the 
        <a
          class="link btn btn-link text-light font-weight-bold textLight p-0 inlineLink"
          href="/docs/topics/repositories"
          rel="noopener noreferrer"
          role="button"
          target="_blank"
        >
          repositories guide
        </a>
         for more information about how to list your content on Artifact Hub.
      </div>
    </div>
    <div />
    <div
      class="py-5 textLight about"
    >
      <div
        class="container-lg px-4 px-sm-0 py-0 py-md-5"
      >
        <div
          class="text-center px-3 px-xs-0"
        >
          Artifact Hub is a web-based application that enables finding, installing, and publishing packages and configurations for CNCF projects. For example, this could include Helm charts and plugins, Falco configurations, Open Policy Agent (OPA) policies, OLM operators, Tinkerbell actions, kubectl plugins, Tekton tasks, KEDA scalers and CoreDNS plugins.
          <div
            class="mx-0 mx-md-3 mx-lg-5 my-4 my-sm-5 d-flex flex-row align-items-strecht justify-content-around"
          >
            <a
              class="link col iconLink"
              href="https://helm.sh"
              rel="noopener noreferrer"
              role="button"
              target="_blank"
            >
              <div
                class="d-flex flex-column justify-content-between align-items-center h-100"
              >
                <img
                  alt="Icon"
                  class="aboutIcon"
                  src="/static/media/helm-chart-light.svg"
                />
                <div
                  class="d-none d-sm-block text-light legendIcon"
                >
                  <small
                    class="text-nowrap"
                  >
                    Helm charts and plugins
                  </small>
                </div>
              </div>
            </a>
            <a
              class="link col iconLink"
              href="https://falco.org"
              rel="noopener noreferrer"
              role="button"
              target="_blank"
            >
              <div
                class="d-flex flex-column justify-content-between align-items-center h-100"
              >
                <img
                  alt="Icon"
                  class="aboutIcon"
                  src="/static/media/falco-rules-light.svg"
                />
                <div
                  class="d-none d-sm-block text-light legendIcon"
                >
                  <small>
                    Falco rules
                  </small>
                </div>
              </div>
            </a>
            <a
              class="link col iconLink"
              href="https://www.openpolicyagent.org"
              rel="noopener noreferrer"
              role="button"
              target="_blank"
            >
              <div
                class="d-flex flex-column justify-content-between align-items-center h-100"
              >
                <img
                  alt="Icon"
                  class="aboutIcon"
                  src="/static/media/opa-policies-light.svg"
                />
                <div
                  class="d-none d-sm-block text-light legendIcon"
                >
                  <small>
                    OPA policies
                  </small>
                </div>
              </div>
            </a>
          </div>
          <div
            class="mx-0 mx-md-3 mx-lg-5 my-4 my-sm-5 d-flex flex-row align-items-strecht justify-content-around"
          >
            <a
              class="link col iconLink"
              href="https://github.com/operator-framework"
              rel="noopener noreferrer"
              role="button"
              target="_blank"
            >
              <div
                class="d-flex flex-column justify-content-between align-items-center h-100"
              >
                <img
                  alt="Icon"
                  class="aboutIcon"
                  src="/static/media/olm-operators-light.svg"
                />
                <div
                  class="d-none d-sm-block text-light legendIcon"
                >
                  <small>
                    OLM operators
                  </small>
                </div>
              </div>
            </a>
            <a
              class="link col iconLink"
              href="https://tinkerbell.org"
              rel="noopener noreferrer"
              role="button"
              target="_blank"
            >
              <div
                class="d-flex flex-column justify-content-between align-items-center h-100"
              >
                <img
                  alt="Icon"
                  class="aboutIcon"
                  src="/static/media/tinkerbell-actions-light.svg"
                />
                <div
                  class="d-none d-sm-block text-light legendIcon"
                >
                  <small>
                    Tinkerbell actions
                  </small>
                </div>
              </div>
            </a>
            <a
              class="link col iconLink"
              href="https://krew.sigs.k8s.io"
              rel="noopener noreferrer"
              role="button"
              target="_blank"
            >
              <div
                class="d-flex flex-column justify-content-between align-items-center h-100"
              >
                <img
                  alt="Icon"
                  class="aboutIcon"
                  src="/static/media/krew-plugins-light.svg"
                />
                <div
                  class="d-none d-sm-block text-light legendIcon"
                >
                  <small>
                    Kubectl plugins
                  </small>
                </div>
              </div>
            </a>
          </div>
          <div
            class="mx-0 mx-md-3 mx-lg-5 my-4 my-sm-5 d-flex flex-row align-items-strecht justify-content-around"
          >
            <a
              class="link col iconLink"
              href="https://tekton.dev"
              rel="noopener noreferrer"
              role="button"
              target="_blank"
            >
              <div
                class="d-flex flex-column justify-content-between align-items-center h-100"
              >
                <img
                  alt="Icon"
                  class="aboutIcon"
                  src="/static/media/tekton-pkg-light.svg"
                />
                <div
                  class="d-none d-sm-block text-light legendIcon"
                >
                  <small>
                    Tekton tasks
                  </small>
                </div>
              </div>
            </a>
            <a
              class="link col iconLink"
              href="https://keda.sh"
              rel="noopener noreferrer"
              role="button"
              target="_blank"
            >
              <div
                class="d-flex flex-column justify-content-between align-items-center h-100"
              >
                <img
                  alt="Icon"
                  class="aboutIcon"
                  src="/static/media/keda-scaler-light.svg"
                />
                <div
                  class="d-none d-sm-block text-light legendIcon"
                >
                  <small>
                    KEDA scalers
                  </small>
                </div>
              </div>
            </a>
            <a
              class="link col iconLink"
              href="https://coredns.io"
              rel="noopener noreferrer"
              role="button"
              target="_blank"
            >
              <div
                class="d-flex flex-column justify-content-between align-items-center h-100"
              >
                <img
                  alt="Icon"
                  class="aboutIcon"
                  src="/static/media/coredns-plugin-light.svg"
                />
                <div
                  class="d-none d-sm-block text-light legendIcon"
                >
                  <small>
                    CoreDNS plugins
                  </small>
                </div>
              </div>
            </a>
          </div>
          Discovering artifacts to use with CNCF projects can be difficult. If every CNCF project that needs to share artifacts creates its own Hub this creates a fair amount of repeat work for each project and a fractured experience for those trying to find the artifacts to consume. The Artifact Hub attempts to solve that by providing a single experience for consumers that any CNCF project can leverage.
        </div>
      </div>
    </div>
    <div
      class="py-5 text-secondary extraInfo"
    >
      <div
        class="container-lg px-sm-4 px-lg-0 py-0 py-md-5"
      >
        <div
          class="text-center px-4 px-xs-0"
        >
          <img
            alt="Logo CNCF sandbox project"
            class="logo homeLogo"
            src="/static/media/cncf-sandbox-horizontal-color.png"
          />
          <div
            class="px-3 pt-4"
          >
            Artifact Hub is a 
            <a
              class="link font-weight-bold text-primary"
              href="https://www.cncf.io/sandbox-projects/"
              rel="noopener noreferrer"
              role="button"
              target="_blank"
            >
              Cloud Native Computing Foundation
            </a>
             sandbox project.
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
